checker/src/main/java/org/checkerframework/checker/optional/OptionalTransfer.java

```diff
@@ -98,4 +149,107 @@ public CFStore initialStore(UnderlyingAST underlyingAST, List<LocalVariableNode>
 
     return result;
   }
+
+  @Override
+  public TransferResult<CFValue, CFStore> visitMethodInvocation(
+      MethodInvocationNode n, TransferInput<CFValue, CFStore> in) {
+    TransferResult<CFValue, CFStore> result = super.visitMethodInvocation(n, in);
+    if (n.getTree() == null) {
+      return result;
+    }
+    return refineNonEmptyToPresentStreamResult(n, result);
+  }
+
+  /**
+   * Refines the result of a call to a method in {@link #nonEmptyToPresentStreamMethods}. Examples
+   * are {@link java.util.stream.Stream#max(Comparator)} and {@link
+   * java.util.stream.Stream#reduce(BinaryOperator)}.
+   *
+   * <p>When one of those methods is invoked on an empty stream, the result is an empty/absent
+   * Optional. When one of those methods is invoked on a non-empty stream, the result is a present
+   * Optional.
+   *
+   * @param n the method invocation node
+   * @param result the transfer result to side effect
+   * @return the refined transfer result
+   */
+  private TransferResult<CFValue, CFStore> refineNonEmptyToPresentStreamResult(
+      MethodInvocationNode methodInvok, TransferResult<CFValue, CFStore> result) {
+    if (NodeUtils.isMethodInvocation(
+        methodInvok, nonEmptyToPresentStreamMethods, optionalTypeFactory.getProcessingEnv())) {
+      if (isReceiverParameterNonEmpty(methodInvok)) {
+        // The receiver is @Non-Empty, therefore the result is @Present.
+        JavaExpression methodInvokJE = JavaExpression.fromNode(methodInvok);
+        if (assumeDeterministic) {
+          insertIntoStoresPermitNonDeterministic(result, methodInvokJE, PRESENT);
+        } else {
+          insertIntoStores(result, methodInvokJE, PRESENT);
+        }
+        CFValue value = analysis.createSingleAnnotationValue(PRESENT, methodInvok.getType());
+
+        return new ConditionalTransferResult<>(
+            finishValue(value, result.getThenStore(), result.getElseStore()),
+            result.getThenStore(),
+            result.getElseStore(),
+            result.getExceptionalStores());
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Returns true if the receiver parameter of the method being invoked is explicitly annotated
+   * with @{@link NonEmpty}.
+   *
+   * @param methodInvok a method invocation node
+   * @return true if the receiver parameter of the method being invoked is explicitly annotated
+   *     with @{@link NonEmpty}
+   */
+  private boolean isReceiverParameterNonEmpty(MethodInvocationNode methodInvok) {
+    ExpressionTree receiverTree = TreeUtils.getReceiverTree(methodInvok.getTree());
+    JavaExpression receiver;
+    if (receiverTree instanceof MethodInvocationTree) {
+      receiver = JavaExpression.getLeftmostReceiverOfMethodInvocation(receiverTree);
```

Comment:
@jyoo980 Ping on this comment.

Reply:
I am sorry for neglecting this comment, thanks for your patience!

You are correct. The following test case fails under the current implementation

```java
import java.util.Optional;
import java.util.List;

import org.checkerframework.checker.nonempty.qual.*;

class Main {

  class Container {

    @NonEmpty List<String> getStrs() {
      throw new Error();
    }
  }

  void foo() {
    Container container = new Container();
    container.getStrs().stream().findAny().get();
  }
}
```

> What is wanted is the leftmost receiver, such that all methods to the right of it are `@PolyNonEmpty`.

I think I agree with the general approach. But, in this case, wouldn't the receiver in question be `container`? Since `getStrs()` is annotated with `@NonEmpty`.

I think what we want instead is the _first_ leftmost receiver that has type `@PolyNonEmpty`, but perhaps I'm mistaken.

Reply:
Please go ahead and implement and document the behavior, and I will take a look.  Thanks!